go/vumitools/contact/models.py

```diff
@@ -381,3 +394,80 @@ def contact_for_addr(self, delivery_class, addr, create=True):
                 "Contact with address '%s' for delivery class '%s' not found."
                 % (addr, delivery_class))
         returnValue(contact)
+
+
+class PaginatedSearch(object):
+    """
+    This has the same external interface as an IndexPage object, but it
+    performs the search queries itself internally.
+
+    To avoid having to perform the first search query externally, the first
+    page of results can be acquired by passing in an empty results list and a
+    cursor of ``0``.
+    """
+
+    def __init__(self, model_proxy, max_results, query, cursor, results):
+        self._model_proxy = model_proxy
+        self.manager = model_proxy._manager
+        self._max_results = max_results
+        self._query = query
+        self._cursor = cursor
+        self._results = results
+
+    @Manager.calls_manager
+    def _get_page_of_results(self):
+        results = yield self._model_proxy.real_search(
+            self._query, rows=self._max_results, start=self._cursor)
+        if len(results) == 0:
+            new_cursor = None
+        else:
+            new_cursor = self._cursor + len(results)
+        returnValue((new_cursor, results))
+
+    def __iter__(self):
+        return iter(self._results)
+
+    def has_next_page(self):
+        return self._cursor is not None
+
+    @Manager.calls_manager
+    def next_page(self):
+        if self._cursor is None:
+            returnValue(None)
+        cursor, results = yield self._get_page_of_results()
+        returnValue(type(self)(
+            self._model_proxy, self._max_results, self._query, cursor,
+            results))
+
+
+class ChainedIndexPages(object):
+    """
+    Wrapper around a list of index pages to walk through them one after the
+    other.
+
+    NOTE: This assumes that all index pages are non-None.
+    """
+
+    def __init__(self, manager, current_page, *further_pages):
+        self.manager = manager
+        self._current_page = current_page
+        self._further_pages = further_pages
+
+    def __iter__(self):
+        return iter(self._current_page)
+
+    def has_next_page(self):
+        if self._current_page.has_next_page():
+            return True
+        return len(self._further_pages) > 0
+
+    @Manager.calls_manager
+    def next_page(self):
+        pages = list(self._further_pages[:])
+        next_current_page = yield self._current_page.next_page()
+        if next_current_page is not None:
+            pages = [next_current_page] + pages
+        if pages:
+            returnValue(type(self)(self.manager, *pages))
+        else:
+            returnValue(None)
```

Comment:
I would think that the way `PaginatedSearch` is used would look the same regardless of how the searching it does looks under the hood. I'm not convinced, but my request is bikesheddy enough to leave as is for now.